src/flint/FlintTypes.jl

```diff
@@ -5739,72 +5713,54 @@ mutable struct FqPolyRepMatrix <: MatElem{FqPolyRepFieldElem}
 
   function FqPolyRepMatrix(r::Int, c::Int, arr::AbstractMatrix{FqPolyRepFieldElem}, ctx::FqPolyRepField)
     z = FqPolyRepMatrix(r, c, ctx)
-    GC.@preserve z for i = 1:r
-      for j = 1:c
-        ccall((:fq_mat_entry_set, libflint), Nothing,
-              (Ref{FqPolyRepMatrix}, Int, Int, Ref{FqPolyRepFieldElem}, Ref{FqPolyRepField}),
-              z, i - 1, j - 1, arr[i, j], ctx)
-      end
+    for i = 1:r, j = 1:c
+      el = arr[i, j]
+      @inbounds z[i, j] = el
     end
     return z
   end
 
   function FqPolyRepMatrix(r::Int, c::Int, arr::AbstractVector{FqPolyRepFieldElem}, ctx::FqPolyRepField)
     z = FqPolyRepMatrix(r, c, ctx)
-    GC.@preserve z for i = 1:r
-      for j = 1:c
-        ccall((:fq_mat_entry_set, libflint), Nothing,
-              (Ref{FqPolyRepMatrix}, Int, Int, Ref{FqPolyRepFieldElem}, Ref{FqPolyRepField}),
-              z, i - 1, j - 1, arr[(i - 1) * c + j], ctx)
-      end
+    for i = 1:r, j = 1:c
+      el = arr[(i - 1) * c + j]
+      @inbounds z[i, j] = el
     end
     return z
   end
 
   function FqPolyRepMatrix(r::Int, c::Int, arr::AbstractMatrix{ZZRingElem}, ctx::FqPolyRepField)
```

Comment:
We could merge three constructors in one now:
```suggestion
  function FqPolyRepMatrix(r::Int, c::Int, arr::AbstractMatrix{<:Union{FqPolyRepFieldElem, ZZRingElem, Integer}}, ctx::FqPolyRepField)
```

But taking a step back, this is a strange constructor: why does it take a matrix *and* dimensions? They are only ever called from a tiny set of places anyway:
```
function (a::FqPolyRepMatrixSpace)(arr::AbstractVector{T}) where {T <: Integer}
  _check_dim(nrows(a), ncols(a), arr)
  return FqPolyRepMatrix(nrows(a), ncols(a), arr, base_ring(a))
end

...

function (a::FqPolyRepMatrixSpace)(arr::AbstractMatrix{ZZRingElem})
  _check_dim(nrows(a), ncols(a), arr)
  return FqPolyRepMatrix(nrows(a), ncols(a), arr, base_ring(a))
end

...

function matrix(R::FqPolyRepField, arr::AbstractMatrix{<: Union{FqPolyRepFieldElem, ZZRingElem, Integer}})
  z = FqPolyRepMatrix(size(arr, 1), size(arr, 2), arr, R)
  return z
end

function matrix(R::FqPolyRepField, r::Int, c::Int, arr::AbstractVector{<: Union{FqPolyRepFieldElem, ZZRingElem, Integer}})
  _check_dim(r, c, arr)
  z = FqPolyRepMatrix(r, c, arr, R)
  return z
end

```

So we could easil drop the `r` and `c` arguments for those which take a matrix. Then we can actually write `@inbounds z[i, j] = arr[i, j]` as the bounds would be valid for both matrices.

Reply:
sounds sensible

Reply:
please have a look again